### PR TITLE
Database Model Consistency Issues

### PR DESCRIPTION
**Pull request by patrick-du**
## Description
- Updated all instances of `teamID` and `userID` to camelcase for consistency
- Renamed `Orders` model to `Order` (https://sequelize.org/master/manual/naming-strategies.html#singular-vs--plural)
- Make User model ID a UUID
- Abstracted unrelated methods in `webhooks.js` into own controllers (`controllers/order.js`, `controllers/team.js`) 
- Added DB table associations
- Current state of DB tables is:
image

Link to ticket: https://www.notion.so/uwblueprintexecs/Building-Up-be1b30c321554bffba37a64c6cdc1ec6?p=b8c768c0863e47a5a8222ce5e33f66a2

### CODE REVIEW

File: src/models/index.js [hunk 1]
```diff
@@ -30,6 +30,5 @@ Object.keys(db).forEach(modelName => {
 });
 
 db.sequelize = sequelize;
-db.Sequelize = Sequelize;
```
Comment (patrick-du) [inline src/models/index.js]: This export of the library `Sequelize` doesn't seem to be used anywhere. Alternatively, in other places, we can access it through `require('sequelize')`. Anyone know if there was a specific reason we had this?

**Comment by patrick-du** [comment]
To address this point in the ticket:

> Can we define the DB table structure in like a js file that's not a "migration" file? maybe as part of the model? or is defining the details in the migration the correct pattern to follow

From the documentation, models are an abstraction of the table in the DB (https://sequelize.org/master/manual/model-basics.html#concept). Aside from user-defined attributes, Sequelize will automatically add `id`, `createdAt`, and `updatedAt` fields by default. With this in mind, we can refer to the model as our source of truth for DB table structure. Another way to keep track could be to use an entity-relationship diagram that we keep up to date.

Migrations are the correct pattern to follow - alternatively, there is model synchronization which ensures tables with the correct columns are automatically created if they don't already exist, however, it's not recommended (https://sequelize.org/master/manual/model-basics.html#synchronization-in-production).

I haven't used Sequelize too much so if the explanation is off, please let me know 😃

File: src/models/order.js [hunk 2]
```diff
@@ -11,20 +11,20 @@ module.exports = (sequelize, DataTypes) => {
       // define association here
     }
   }
-  Orders.init(
+  Order.init(
     {
       orderNumber: DataTypes.INTEGER,
```
Comment (patrick-du) [inline src/models/order.js:16]: This seems redundant to `id`. I was thinking that an entry for Order is only created when a user places it so we can replace this with `id`.

File: src/controllers/webhooks.js [hunk 3]
```diff
@@ -1,27 +1,20 @@
...
 };
 
 function findById(orderNumber) {
-  return models.Orders.findOne({
+  return models.Order.findOne({
     where: {
       orderNumber,
     },
   });
 }
```
Comment (patrick-du) [inline src/controllers/webhooks.js:16]: If we replace the usage of `orderNumber` with `id` in our model, we can replace this with `models.Order.findByPk()`
https://sequelize.org/master/manual/model-querying-finders.html#-code-findbypk--code-

File: src/controllers/webhooks.js [hunk 4]
```diff
@@ -109,46 +92,42 @@ const captureOrderWebhook = async (req, res) => {
...
+        teamId: noteAttributesMap.get(noteAttributesEnum.teamId),
+        price: subtotal_price,
         numberOfItems,
-        purchaseDate,
+        purchaseDate: created_at,
       });
 
-      // succesful payment record creation, now we can add to redis.
-      incrementTeamScore(parseInt(noteAttributesMap.get(noteAttributesEnum.teamID), 10), numberOfItems);
+      incrementTeamScore(parseInt(noteAttributesMap.get(noteAttributesEnum.teamId), 10), numberOfItems);
```
Comment (ZhangCreations) [inline src/controllers/webhooks.js]: Maybe we should change this function name to something like "incrementTeamSales" or something since we're no longer using the redis scoring method?

**Review by ZhangCreations — APPROVED** [review]

File: src/controllers/webhooks.js [hunk 5]
```diff
@@ -72,26 +57,24 @@ function computeUpdatedPrice(event) {
 function orderChanges(incomingOrder, existingOrder) {
   // this function will return an object of updated values relevant to our DB table
   const changelog = {};
+  const { note_attributes } = incomingOrder;
+  const { userId, teamId, price, numberOfItems } = existingOrder;
```
Comment (jzxhuang) [inline src/controllers/webhooks.js:61]: 🎉

File: src/controllers/webhooks.js [hunk 6]
```diff
@@ -109,46 +92,42 @@ const captureOrderWebhook = async (req, res) => {
...
     console.error(`Error: ${err.message}`);
   }
-  // the Shopify order number from the JSON payload.
-  const orderNumber = event.order_number;
+
+  // order_number - Shopify order number
+  // subtotal_price - includes any discounts, excludes tax
+  // date of purchase from Shopify request data.
+
+  const { order_number, note_attributes, subtotal_price, created_at } = event;
```
Comment (jzxhuang) [inline src/controllers/webhooks.js:100]: 🎉

File: src/models/team.js [hunk 7]
```diff
@@ -8,15 +8,24 @@ module.exports = (sequelize, DataTypes) => {
      * The `models/index` file will call this method automatically.
      */
     static associate(models) {
-      // define association here
+      Team.hasMany(models.User);
+      Team.hasMany(models.Order);
```
Comment (jzxhuang) [inline src/models/team.js:12]: nice

File: src/migrations/20201029230535-create-orders.js [hunk 8]
```diff
@@ -12,30 +12,12 @@ module.exports = {
...
-            schema: 'public',
-          },
-          key: 'id',
-        },
+      userId: {
+        type: Sequelize.UUID,
         allowNull: false,
       },
-      teamID: {
+      teamId: {
```
Comment (jzxhuang) [inline src/migrations/20201029230535-create-orders.js:19]: I think this should also be `UUID`, but you may have not done so because Kevin/Hyunzee's PR will be changing it to `STRING` so no worries

**Review by jzxhuang — COMMENTED** [review]
Nice, thanks for refactoring other pieces of code as you go! And thanks for the explanation on migrations being the correct pattern

Only possible change: Seems like Sequelize model names should be `singular`, what do you think about using singular for the actual postgres table names?

i.e.
```
await queryInterface.createTable('Teams', {
```

becomes
``` 
await queryInterface.createTable('Team', {
```

File: src/models/index.js (hunk 1 above)
Comment (jzxhuang) [reply src/models/index.js]: yeah idk why this exists, I guess we don't need it but it's not a big deal if we leave it for now

**Comment by patrick-du** [comment]
> Only possible change: Seems like Sequelize model names should be `singular`, what do you think about using singular for the actual postgres table names?

https://sequelize.org/master/manual/model-basics.html#table-name-inference

TLDR; Sequelize offers two solutions to table naming:
1. By default, if no table name is provided (our current approach), Sequelize pluralizes the model name and uses that as the table name
2. Enforce table names to be equal to model name by using the option `freezeTableName: true` in model declaration

I'm indifferent to both approaches - what are your thoughts @cynding @jzxhuang?

**Comment by jzxhuang** [comment]
> https://sequelize.org/master/manual/model-basics.html#table-name-inference
> 
> TLDR; Sequelize offers two solutions to table naming:
> 
> 1. By default, if no table name is provided (our current approach), Sequelize pluralizes the model name and uses that as the table name
> 2. Enforce table names to be equal to model name by using the option `freezeTableName: true` in model declaration
> 
> I'm indifferent to both approaches - what are your thoughts @cynding @jzxhuang?

Thanks for linking the docs! Also indifferent — let's use their default approach and ship it

**Review by jzxhuang — APPROVED** [review]
noice

gonna make a [separate ticket](https://www.notion.so/uwblueprintexecs/Building-Up-be1b30c321554bffba37a64c6cdc1ec6?p=e45671aee2a04a98a54c6e8fcf26bf20) to run the migrations on prod after also merging #35
